aspnetcore/blazor/webassembly-lazy-load-assemblies.md

```diff
@@ -0,0 +1,163 @@
+`LazyAssemblyLoader` is a framework-provided singleton service for loading assemblies. Inject `LazyAssemblyLoader` into the `Router` component:
+
+```razor
+...
+@using Microsoft.AspNetCore.Components.WebAssembly.Services
+@inject LazyAssemblyLoader assemblyLoader
+
+...
+```
+
+The `LazyAssemblyLoader` provides the `LoadAssembliesAsync` method that:
+
+* Uses JS interop to fetch assemblies via a network call.
+* Loads assemblies into the runtime executing on WASM in the browser.
+
+> [!NOTE]
+> The framework's lazy loading implementation supports prerendering on the server. During prerendering, all assemblies, including those marked for lazy loading, are assumed to be loaded.
+
+### User interaction with `<Navigating>` content
+
+While loading assemblies, which can take several seconds, the `Router` component can indicate to the user that a page transition is occurring:
+
+* Add an [`@using`](xref:mvc/views/razor#using) directive for the <xref:Microsoft.AspNetCore.Components.Routing?displayProperty=fullName> namespace.
+* Add a `<Navigating>` tag to the component with markup to display during page transition events.
+
+```razor
+...
+@using Microsoft.AspNetCore.Components.Routing
+...
+
+<Router ...>
+    <Navigating>
+        <div style="...">
+            <p>Loading the requested page&hellip;</p>
+        </div>
+    </Navigating>
+</Router>
+
+...
+```
+
+### Complete example
+
+The following complete `Router` component demonstrates loading the `GrantImaharaRobotControls.dll` assembly when the user navigates to `/robot`. During page transitions, a styled message is displayed to the user.
+
+```razor
+@using System.Reflection
+@using Microsoft.AspNetCore.Components.Routing
+@using Microsoft.AspNetCore.Components.WebAssembly.Services
+@inject LazyAssemblyLoader assemblyLoader
+
+<Router AppAssembly="@typeof(Program).Assembly" 
+    AdditionalAssemblies="@lazyLoadedAssemblies" OnNavigateAsync="@OnNavigateAsync">
+    <Navigating>
+        <div style="padding:20px;background-color:blue;color:white">
+            <p>Loading the requested page&hellip;</p>
+        </div>
+    </Navigating>
+    <Found Context="routeData">
+        <RouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)" />
+    </Found>
+    <NotFound>
+        <LayoutView Layout="@typeof(MainLayout)">
+            <p>Sorry, there's nothing at this address.</p>
+        </LayoutView>
+    </NotFound>
+</Router>
+
+@code {
+    private List<Assembly> lazyLoadedAssemblies = new List<Assembly>();
+
+    private async Task OnNavigateAsync(NavigationContext args)
+    {
+        try
+        {
+            if (args.Path.EndsWith("/robot"))
+            {
+                var assemblies = await assemblyLoader.LoadAssembliesAsync(
+                    new List<string>() { "GrantImaharaRobotControls.dll" });
+                lazyLoadedAssemblies.AddRange(assemblies);
+            }
+        }
+        catch (Exception ex)
+        {
+            ...
+        }
+    }
+}
+```
```

Comment:
We should add a note that says if the `OnNavigateAsync` callback throws an unhandled exception then we will surface the Blazor error UI (the little yellow bar at the bottom).

This is a recent change that I am introducing in https://github.com/dotnet/aspnetcore/pull/23947.